internal/cmd/gocdk/serve.go

```diff
@@ -91,47 +128,73 @@ func serveLoop(ctx context.Context, pctx *processContext, logger *log.Logger, ad
 	}()
 
 	// Apply Terraform configuration in biome.
-	if err := apply(ctx, pctx, []string{biome}); err != nil {
+	if err := apply(ctx, pctx, []string{opts.biome}); err != nil {
 		return err
 	}
 
 	// Build and run the server.
-	// TODO(light): Bind host is ignored for now because eventually this
-	// subprocess will be reverse proxied. The proxy should respect the host.
-	_, portString, err := net.SplitHostPort(address)
-	if err != nil {
-		return err
-	}
-	port, err := net.LookupPort("tcp", portString)
-	if err != nil {
-		return xerrors.Errorf("resolve port for %q: %w", address, err)
-	}
-	alloc := &serverAlloc{
-		exePath: filepath.Join(buildDir, "server"),
-		port:    port,
+	allocA := &serverAlloc{
+		exePath: filepath.Join(buildDir, "serverA"),
+		port:    opts.basePort,
 	}
-	logger.Println("Building server...")
-	if err := buildForServe(ctx, pctx, moduleRoot, alloc.exePath); err != nil {
-		return err
+	allocB := &serverAlloc{
+		exePath: filepath.Join(buildDir, "serverB"),
+		port:    opts.basePort + 1,
 	}
-	process, err := alloc.start(ctx, pctx, logger)
-	if err != nil {
-		return err
+	currAlloc, nextAlloc := allocA, allocB
+	var process *exec.Cmd
+loop:
+	for first := true; ; first = false {
+		// After the first iteration of the loop, each iteration should wait for a
+		// change in the filesystem before proceeding.
+		if !first {
+			// TODO(#1881): Actually check filesystem instead of SIGUSR1.
+			select {
+			case <-reload:
+			case <-ctx.Done():
+				break loop
+			}
+		}
+
+		// Build and run the server.
+		logger.Println("Building server...")
+		if err := buildForServe(ctx, pctx, opts.moduleRoot, currAlloc.exePath); err != nil {
+			logger.Printf("Build: %v", err)
+			if process == nil {
+				myProxy.setBuildError(err)
+			}
+			continue
+		}
+		newProcess, err := currAlloc.start(ctx, pctx, logger, opts.moduleRoot)
+		if err != nil {
+			logger.Printf("Starting server: %v", err)
+			if process == nil {
+				myProxy.setBuildError(err)
+			}
+			continue
+		}
+
+		// Server started successfully. Cut traffic over.
+		myProxy.setBackend(currAlloc.url(""))
+		if process == nil {
```

Comment:
could we also use `first` for this?

Reply:
No: the loop could iterate a few times before the server starts successfully in the case of failure to build or failure to start. `process != nil` indicates that a process has built and started successfully in the past.